sys/net/application_layer/gcoap/dns.c

```diff
@@ -113,9 +122,16 @@ static ssize_t _send(const void *buf, size_t len, const sock_udp_ep_t *remote,
 
 int gcoap_dns_query(const char *domain_name, void *addr_out, int family)
 {
+    int res;
+
+    mutex_lock(&_client_mutex);
+    if ((res = dns_cache_query(domain_name, addr_out, family)) > 0) {
+        mutex_unlock(&_client_mutex);
+        return res;
```

Comment:
Usually DNS operations are synchronous, so I don't think this is all that much of an issue. However, `gcoap` works asynchronous, so I had to account for that here.

Reply:
Both DoC and regular DNS could be around using the cache, so the dns_cache would need to be synchronized if it's not thread safe on its own. The _client_mutex of DoC could probably be acquired only after the cache was missed.

Reply:
> Both DoC and regular DNS could be around using the cache, so the dns_cache would need to be synchronized if it's not thread safe on its own.

True! I was just thinking about that and if it would be required, if the source transport of a record must be marked in the cache somehow... so only DoC queried records are fetched for DoC and only DNS queried records are fetched for DNS from the cache. I'm pretty sure it is not necessary, but just as something to ponder :D

> The _client_mutex of DoC could probably be acquired only after the cache was missed.

Yepp, I guess I can move that down, assuming the cache access is synchronized.